### PR TITLE
Move from uglifyjs-webpack-plugin to terser-webpack-plugin

### PR DESCRIPTION
**Pull request by abenezerabebe**
**What kind of change does this PR introduce?**

Replaces `uglifyjs-webpack-plugin` dependency with `terser-webpack-plugin` and refactores any packages that used `uglifyjs-webpack-plugin`.

**Did you add tests for your changes?**
I have updated existing tests and replaced outdated tests with new ones.

**If relevant, did you update the documentation?**
I have updated the documentation for the `migrate` package.

**Summary**

Webpack moved from `uglifyjs-webpack-plugin` to `terser-webpack-plugin` in [#8036](https://github.com/webpack/webpack/pull/8036)
The motivation there was that `uglifyjs-webpack-plugin` uses  `uglify-es` which is no longer being maintained.

Fixes [#701](https://github.com/webpack/webpack-cli/issues/701)

**Does this PR introduce a breaking change?**

None I am aware of.

**Other information**
This is my first contribution! I may have missed some things and am super open to feedback 😄

### CODE REVIEW

File: packages/migrate/terserPlugin/__testfixtures__/terserPlugin-4.input.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
+module.exports = {
```
Comment (EugeneHlushko) [inline packages/migrate/terserPlugin/__testfixtures__/terserPlugin-4.input.js]: this looks misaligned compared to others

File: packages/utils/ast-utils.test.js [hunk 2]
```diff
@@ -34,28 +34,28 @@ describe("utils", () => {
 	describe("findPluginsByName", () => {
 		it("should find plugins in AST", () => {
 			const ast = j(`
-{ foo: new webpack.optimize.UglifyJsPlugin() }
+{ foo: new webpack.optimize.TerserPlugin() }
```
Comment (ematipico) [inline packages/utils/ast-utils.test.js]: Is terser plugin inside optimize? Genuine question, I can't remember

Comment (EugeneHlushko) [reply packages/utils/ast-utils.test.js]: hmm i think its in separate package e.g. `const TerserPlugin = require('terser-webpack-plugin');`

File: packages/migrate/loaderOptionsPlugin/loaderOptionsPlugin.ts [hunk 3]
```diff
@@ -37,8 +37,8 @@ export default function(j: IJSCodeshift, ast: INode): INode {
 			});
 	}
 
-	// If there is UglifyJsPlugin, set minimize: true
-	if (findPluginsByName(j, ast, ["webpack.optimize.UglifyJsPlugin"]).size()) {
+	// If there is TerserPlugin, set minimize: true
+	if (findPluginsByName(j, ast, ["webpack.optimize.TerserPlugin"]).size()) {
```
Comment (PlayMa256) [inline packages/migrate/loaderOptionsPlugin/loaderOptionsPlugin.ts]: webpack has no export on `webpack.optimise.TerserPlugin`. The old code has to be kept. We want to move from `webpack.optimise.UglifyJsPlugin` on `plugin[]` to `optimize.minimizer[]` then add terser.

File: packages/migrate/__snapshots__/migrate.test.js.snap [hunk 4]
```diff
@@ -22,7 +22,7 @@ module.exports = {
 		modules: ['node_modules']
 	},
 	plugins: [
-		new webpack.optimize.UglifyJsPlugin(),
+		new webpack.optimize.TerserPlugin(),
```
Comment (evenstensberg) [inline packages/migrate/__snapshots__/migrate.test.js.snap]: terser plugin and you will need to add require statement somewhere

File: packages/migrate/loaderOptionsPlugin/loaderOptionsPlugin.ts (hunk 3 above)
Comment (evenstensberg) [reply packages/migrate/loaderOptionsPlugin/loaderOptionsPlugin.ts]: yes. So @abenezerabebe you will need to add a require statement and use `TerserPlugin()` instead

Comment (abenezerabebe) [reply packages/migrate/loaderOptionsPlugin/loaderOptionsPlugin.ts]: @evenstensberg I've replace lookups for `TerserPlugin` in `webpack.optimise` with requiring `terser-webpack-plugin` and using TerserPlugin itself. Do I need to do anything concerning `optimize.minimizer[]`?

**Comment by abenezerabebe** [comment]
@evenstensberg I misunderstood, thanks for taking the time to clarify! What I understand so far is:
- search the ast for a require for `UglifyJsPlugin`
- remove it and add require for `TerserJsPlugin`, `terser-webpack-plugin`
- reset the files back to uglify
Is that right?

Lastly, it looks to me like the transformarion removes the plugin from plugins[] and adds it to minimizer[], is that correct?

**Comment by PlayMa256** [comment]
> 
> 
> @evenstensberg I misunderstood, thanks for taking the time to clarify! What I understand so far is:
> 
>     * search the ast for a require for `UglifyJsPlugin`
> 
>     * remove it and add require for `TerserJsPlugin`, `terser-webpack-plugin`
> 
>     * reset the files back to uglify
>       Is that right?
> 
> 
> Lastly, it looks to me like the transformarion removes the plugin from plugins[] and adds it to minimizer[], is that correct?

Kind of.

- Search for `webpack.optimise.UglifyPlugin` OR `UglifyPlugin` (here the second one can have any name, better get the variable name that was declared on the require statement).

- If any found, add `optimizations.minimize:true`.

- Remove those from `plugin[]`

- If Uglify had any options, add a require to `terser-webpack-plugin`, add that to `optimizations.minimizer[]` and add the options.

File: packages/migrate/uglifyJsPlugin/__snapshots__/uglifyJsPlugin.test.js.snap [hunk 5]
```diff
@@ -78,7 +78,7 @@ exports[`uglifyJsPlugin transforms correctly using "uglifyJsPlugin-3" data 1`] =
 `;
 
 exports[`uglifyJsPlugin transforms correctly using "uglifyJsPlugin-4" data 1`] = `
-"const UglifyJsPlugin = require('uglifyjs-webpack-plugin');
+"const TerserPlugin = require('terser-webpack-plugin');
```
Comment (PlayMa256) [inline packages/migrate/uglifyJsPlugin/__snapshots__/uglifyJsPlugin.test.js.snap]: Interesting, why not all snapshots were updated? Just these 2?

Comment (abenezerabebe) [reply packages/migrate/uglifyJsPlugin/__snapshots__/uglifyJsPlugin.test.js.snap]: @PlayMa256 Good question! This is the only one that failed during test run

**Comment by abenezerabebe** [comment]
Locally all tests are passing but the CI's are unhappy. Any clues on what might be going on?

**Comment by evenstensberg** [comment]
Can u update against master branch and update snapshots?

**Comment by abenezerabebe** [comment]
> Can u update against master branch and update snapshots?

@evenstensberg This branch is up to date with `master`

I'm not sure why but I have no failures locally so I can't update the snapshots. Any idea what's going on?

**Comment by EvsChen** [comment]
@abenezerabebe It's because our Travis script does not build the TS. I'm working on a PR #735 for this

**Comment by abenezerabebe** [comment]
> As mentioned, you will need to update the transformations as well, not only the names. Great job so far!

Which transformations need to be updated?

**Comment by evenstensberg** [comment]
Hey, what's the status here?

**Comment by abenezerabebe** [comment]
> Hey, what's the status here?

Hey, I've made changes that update the snapshots but I'm stuck on the transformations that @evenstensberg mentioned need to be updated. I would really like to understand that a bit more!

**Comment by PlayMa256** [comment]
The migrate transformation looks good to me!

**Comment by abenezerabebe** [comment]
I've done some clean up and now I'm stuck on what's missing. Would be great to get feedback on that!

**Comment by evenstensberg** [comment]
What’s missing

**Comment by abenezerabebe** [comment]
> What’s missing

@evenstensberg I don't know if anything is missing. If we're done then let's close this thing!

**Comment by evenstensberg** [comment]
Tersser goes directly to the plugin prop, not in minimize

**Comment by abenezerabebe** [comment]
> Tersser goes directly to the plugin prop, not in minimize

@evenstensberg after the transform, `Terser` should be instantiated in `plugins` instead of `optimization.minimizer`?

I was looking at the docs [here](https://github.com/webpack-contrib/terser-webpack-plugin#getting-started) and it looks like `Terser` is instantiated in `optimization.minimizer`

**Comment by PlayMa256** [comment]
> > Tersser goes directly to the plugin prop, not in minimize
> 
> @evenstensberg after the transform, `Terser` should be instantiated in `plugins` instead of `optimization.minimizer`?
> 
> I was looking at the docs [here](https://github.com/webpack-contrib/terser-webpack-plugin#getting-started) and it looks like `Terser` is instantiated in `optimization.minimizer`

Yes, you are right. Terser should go on `optimization.minimizer`.

File: packages/generators/utils/tooltip.ts [hunk 6]
```diff
@@ -52,10 +52,10 @@ export default {
 
 	uglify: (_?: void): string => {
```
Comment (rishabh3112) [inline packages/generators/utils/tooltip.ts]:
```suggestion
	terser: (_?: void): string => {
```

File: packages/generators/init-generator.ts [hunk 7]
```diff
@@ -425,12 +425,12 @@ export default class InitGenerator extends Generator {
 	public installPlugins() {
 		if (this.isProd) {
 			this.dependencies = this.dependencies.filter(
-				(p: string) => p !== "uglifyjs-webpack-plugin",
+				(p: string) => p !== "terser-webpack-plugin",
 			);
 		} else {
 			this.configuration.config.topScope.push(
 				tooltip.uglify(),
```
Comment (rishabh3112) [inline packages/generators/init-generator.ts]:
```suggestion
				tooltip.terser(),
```

File: packages/generators/utils/tooltip.ts (hunk 6 above)
Comment (abenezerabebe) [reply packages/generators/utils/tooltip.ts]: > Looks good to me 👍
> Left few suggestions :)

@rishabh3112 awesome, made the changes!

**Comment by sendilkumarn** [comment]
ping @evenstensberg

**Comment by evenstensberg** [comment]
Let's take this before the migration PR @sendilkumarn

**Comment by sendilkumarn** [comment]
Closing this in favor of #792 

@abenezerabebe Thanks for the PR. ❤️

**Pull request closed without merging**